File: day19_part1_2018.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class solution {

    public static void main(String[] args) {
        try {
            BufferedReader reader = new BufferedReader(new FileReader("input.txt"));
            int ipBind = 0;
            String line = reader.readLine();
            ipBind = Integer.parseInt(line.substring(4));

            String[][] instructions = new String[1000][4];
            int index = 0;
            while ((line = reader.readLine()) != null) {
                instructions[index] = line.split(" ");
                index++;
            }

            int[] registers = new int[6];
            for (int ip = 0; ip < index; ip++) {
                registers[ipBind] = ip;
                String[] inst = instructions[ip];
                String opcode = inst[0];
                int a = Integer.parseInt(inst[1]);
                int b = Integer.parseInt(inst[2]);
                int c = Integer.parseInt(inst[3]);

                switch (opcode) {
                    case "addr":
                        registers[c] = registers[a] + registers[b];
                        break;
                    case "addi":
                        registers[c] = registers[a] + b;
                        break;
                    case "mulr":
                        registers[c] = registers[a] * registers[b];
                        break;
                    case "muli":
                        registers[c] = registers[a] * b;
                        break;
                    case "banr":
                        registers[c] = registers[a] & registers[b];
                        break;
                    case "bani":
                        registers[c] = registers[a] & b;
                        break;
                    case "borr":
                        registers[c] = registers[a] | registers[b];
                        break;
                    case "bori":
                        registers[c] = registers[a] | b;
                        break;
                    case "setr":
                        registers[c] = registers[a];
                        break;
                    case "seti":
                        registers[c] = a;
                        break;
                    case "gtir":
                        registers[c] = boolToInt(a > registers[b]);
                        break;
                    case "gtri":
                        registers[c] = boolToInt(registers[a] > b);
                        break;
                    case "gtrr":
                        registers[c] = boolToInt(registers[a] > registers[b]);
                        break;
                    case "eqir":
                        registers[c] = boolToInt(a == registers[b]);
                        break;
                    case "eqri":
                        registers[c] = boolToInt(registers[a] == b);
                        break;
                    case "eqrr":
                        registers[c] = boolToInt(registers[a] == registers[b]);
                        break;
                }

                ip = registers[ipBind];
                if (ip < 0 || ip >= index) {
                    break;
                }
            }

            System.out.println(registers[0]);

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static int boolToInt(boolean b) {
        if (b) {
            return 1;
        } else {
            return 0;
        }
    }
}

File: objectivec/day19_part1_2018.m
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *fileContents = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:nil];
        NSArray *lines = [fileContents componentsSeparatedByString:@"\n"];
        
        int ipBind = 0;
        sscanf([lines[0] UTF8String], "#ip %d", &ipBind);
        
        NSMutableArray<NSArray<NSString *> *> *instructions = [NSMutableArray array];
        for (int i = 1; i < lines.count; i++) {
            [instructions addObject:[lines[i] componentsSeparatedByString:@" "]];
        }
        
        int registers[6] = {0};
        for (int ip = 0; ip < instructions.count; ip++) {
            registers[ipBind] = ip;
            NSArray<NSString *> *inst = instructions[ip];
            NSString *opcode = inst[0];
            int a = [inst[1] intValue];
            int b = [inst[2] intValue];
            int c = [inst[3] intValue];
            
            if ([opcode isEqualToString:@"addr"]) {
                registers[c] = registers[a] + registers[b];
            } else if ([opcode isEqualToString:@"addi"]) {
                registers[c] = registers[a] + b;
            } else if ([opcode isEqualToString:@"mulr"]) {
                registers[c] = registers[a] * registers[b];
            } else if ([opcode isEqualToString:@"muli"]) {
                registers[c] = registers[a] * b;
            } else if ([opcode isEqualToString:@"banr"]) {
                registers[c] = registers[a] & registers[b];
            } else if ([opcode isEqualToString:@"bani"]) {
                registers[c] = registers[a] & b;
            } else if ([opcode isEqualToString:@"borr"]) {
                registers[c] = registers[a] | registers[b];
            } else if ([opcode isEqualToString:@"bori"]) {
                registers[c] = registers[a] | b;
            } else if ([opcode isEqualToString:@"setr"]) {
                registers[c] = registers[a];
            } else if ([opcode isEqualToString:@"seti"]) {
                registers[c] = a;
            } else if ([opcode isEqualToString:@"gtir"]) {
                registers[c] = (a > registers[b]) ? 1 : 0;
            } else if ([opcode isEqualToString:@"gtri"]) {
                registers[c] = (registers[a] > b) ? 1 : 0;
            } else if ([opcode isEqualToString:@"gtrr"]) {
                registers[c] = (registers[a] > registers[b]) ? 1 : 0;
            } else if ([opcode isEqualToString:@"eqir"]) {
                registers[c] = (a == registers[b]) ? 1 : 0;
            } else if ([opcode isEqualToString:@"eqri"]) {
                registers[c] = (registers[a] == b) ? 1 : 0;
            } else if ([opcode isEqualToString:@"eqrr"]) {
                registers[c] = (registers[a] == registers[b]) ? 1 : 0;
            }
            
            ip = registers[ipBind];
            if (ip < 0 || ip >= instructions.count) {
                break;
            }
        }
        
        printf("%d\n", registers[0]);
    }
    return 0;
}